### DISCUSSION TITLE
Advice on deployment process

### DISCUSSION THREAD

**Discussion by RoyalVICTOR-pro** [adonisjs/core#4283] [Help]
Hello everyone!

I'm looking to deploy an AdonisJS project (on "The World" hosting at Planet Hoster - shared hosting).

I would like to find a good deployment process. I want it to be smooth while favoring stability. Knowing that my project is on GitLab, that when I deploy, I will probably have unit or functional tests, migrations, the build, I'm wondering about the best way to do this. Attached is a draft of what I was initially thinking.

![image](https://github.com/adonisjs/core/assets/145127368/290e89f6-b523-48e6-bb37-4ba1e228655a)

I would prefer to use "Git pull" to update my files on the production server, but after, the 'node ace build --production' command doesn't work on the server (build command doesn't exist). And I don't know how to run my tests on the server either (test command doesn't exist). I don't think I have the possibility to do CI/CD on the server.

If anyone has advice, a resource (other than the documentation), I'm interested. I'm far from being a deployment specialist, don't hesitate if I say big mistakes or if there are simpler things.

Thanks in advance!

**Comment by MaximeMRF** [comment]
Heyy @RoyalVICTOR-pro

I personally use github actions to test and deploy my application but you can do the same with gitlab.

I have two yml files, one for executing tests that is triggered by push and pull request, and one another that build my adonisjs application on a docker image and push it on a registry (triggered only by push).
You have different choice of docker image registry like github (ghcr.io), gitlab or also docker hub.

When my image is pushed on the registry my action login into my server with ssh and log into the registry, pull the new image and replace the old container with a container with the new image version.

Here is a [example](https://github.com/Hexa-ai/hexa-data/blob/master/.github/workflows/adonisjs.yml) but it's quit different of my process.

**Reply by RoyalVICTOR-pro** [reply]
Hi Maxime ! 
Unfortunately I can't have docker on my hosting server because it's shared hosting but thank you for your answer.

**Reply by RoyalVICTOR-pro** [reply]
Regarding the GitLab part, I'm currently facing a lot of issues deploying an AdonisJS application online. The npm installs are not working correctly.

At the hosting provider Planet Hoster, I tried to run npm install, but the presence of the phc-argon2 package prevents me from completing npm install. If I remove it, npm install completes.  But after that, I can't execute node ace build because the 'build' command doesn't exist. This is because the @adonisjs/assembler package is missing. It should be installed, but it isn't. 
If I install the package manually, it says everything is up to date, but when I run node ace --version, it indicates that the assembler is not installed. And when I look in the adonisjs folder in node_modules, the assembler folder is not there. On my local machine, everything's fine !

I tried on another hosting provider. I created a NodeJS application through their interface, git cloned my project into the folder, and ran npm install. It has been stuck for several tens of minutes on 'idealTree:lib: sill idealTree buildDeps', and nothing is happening.

If this sounds familiar or if anyone has an idea, I'm all ears!"

**Reply by MaximeMRF** [reply]
Can you provide your workflow file ? I think it's more simple.

**Reply by RoyalVICTOR-pro** [reply]
For the moment, I connect to the server with the terminal then : 
- I git clone my project
- I go into the folder
- I launch a NPM INSTALL (there I have problems with phc-argon2 package)
- I remove phc-argon2 package from package.json
- I launch a NPM INSTALL (This time, it's ok)
- I launch a NPM RUN BUILD who calls "node ace build" (there I have an error because "build" doesn't exist)
- I try to install @adonisjs/assembler manually with "npm i @adonisjs/assembler" (I have a message Uptodate but there's no assembler folder in node_modules > adonisjs)
- I try a "node ace --version" who said : App version: 1.0.0 │ Framework version: 5.9.0 │ Assembler version: Not Installed

Thank you for your help

**Reply by RoyalVICTOR-pro** [reply]
(And everything is working fine locally)

**Reply by RoyalVICTOR-pro** [reply]
I discover that it was due to the NODE_ENV, by default the NODE_ENV (or APP_ENV) was set to production and in production, the dev dependencies aren't installed.
Thanks for help !

**Reply by MaximeMRF** [reply]
Ah yeah I see, happy to listen that it's working now :)
